deer/fsolve_ivp.py

```diff
@@ -173,3 +174,225 @@ def solve_ivp_inv_lin(self, gmat: List[jnp.ndarray], rhs: jnp.ndarray,
         # compute the recursive matrix multiplication
         yt = matmul_recursive(gtbar, htbar, y0)  # (nt, ny)
         return yt
+
+
+class GeneralODE(SolveIVPMethod):
+    """
+    Compute the solution of initial value problem with the ODE methods.
+    """
+
+    def compute(self, func: Callable[[jnp.ndarray, jnp.ndarray, Any], jnp.ndarray],
+                y0: jnp.ndarray, xinp: jnp.ndarray, params: Any, tpts: jnp.ndarray):
+        
+        # Define the main function to cover all time intervals
+        def scan_step(carry, i):
+            _, success = carry
+            def success_fn(carry, i):
+                yi, _ = carry
+                xi = xinp[i-1]
+                dx = xinp[i] - xinp[i-1]
+                dt = tpts[i] - tpts[i-1]
+                yn, success = self.ode_step(func, yi, xi, dt, dx, params)
+                return yn, success
```

Comment:
Seems like `xinp` and `tpts` variables are coming from outside this function and that would make this function unpure. JAX has a strong obsession with pure function (see [this](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions) for example), so can you make it a pure function?

Reply:
Fixed with [make function purer](https://github.com/machine-discovery/deer/pull/26/commits/da37873b367a24900a6a24876e7559b26c44e76c)